File: sdks/typescript/src/state.ts
/**
 * Global state management for simple API
 */

import { LineageClient } from './client';
import type {
  Actor,
  ActorType,
  Event,
  EventType,
  InitConfig,
  Intent,
  Scope,
} from './types';

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

class GlobalState {
  private initialized = false;
  private client: LineageClient | null = null;
  private scope: Scope | null = null;
  private defaultActor: Actor | null = null;
  private eventTypes: Map<string, EventType> = new Map();
  private actors: Map<string, Actor> = new Map();
  private lastEvent: Event | null = null;
  private waitTime = 2000;

  async init(config: InitConfig): Promise<void> {
    this.client = new LineageClient({ baseUrl: config.baseUrl });
    this.waitTime = (config.waitTime ?? 2) * 1000;

    if (config.autoCreate !== false) {
      // Create scope
      this.scope = await this.client.scopes.create({
        project: config.project,
        domain: config.domain,
        environment: config.environment,
      });

      // Create default actor if specified
      if (config.actorName) {
        this.defaultActor = await this.getOrCreateActor(
          config.actorType || 'service',
          config.actorName
        );
      }
    }

    this.initialized = true;
  }

  private ensureInitialized(): void {
    if (!this.initialized) {
      throw new Error('lineage.init() must be called first');
    }
  }

  private async getOrCreateActor(
    actorType: ActorType,
    actorName: string
  ): Promise<Actor> {
    const key = `${actorType}:${actorName}`;
    const existing = this.actors.get(key);
    if (existing) return existing;

    const actor = await this.client!.actors.create({
      type: actorType,
      name: actorName,
    });
    this.actors.set(key, actor);
    return actor;
  }

  private async getOrCreateEventType(
    name: string,
    allowedIntents?: Intent[]
  ): Promise<EventType> {
    const existing = this.eventTypes.get(name);
    if (existing) return existing;

    const eventType = await this.client!.eventTypes.create({
      name,
      version: '1.0',
      allowed_intents: allowedIntents,
    });
    this.eventTypes.set(name, eventType);
    return eventType;
  }

  private async resolveActor(
    actor?: [ActorType, string] | Actor
  ): Promise<Actor> {
    if (!actor) {
      if (!this.defaultActor) {
        throw new Error('No actor specified and no default actor set');
      }
      return this.defaultActor;
    }

    if (Array.isArray(actor)) {
      return this.getOrCreateActor(actor[0], actor[1]);
    }

    return actor;
  }

  async emit(
    eventType: string,
    intent: Intent,
    payload: Record<string, unknown>,
    options: {
      confidence?: number;
      actor?: [ActorType, string] | Actor;
      parent?: Event | string;
      reason?: string;
      wait?: boolean;
    } = {}
  ): Promise<Event | null> {
    this.ensureInitialized();

    // Resolve actor
    const resolvedActor = await this.resolveActor(options.actor);

    // Get or create event type
    const et = await this.getOrCreateEventType(eventType, [intent]);

    // Resolve parent
    let parentIds: string[] | undefined;
    if (options.parent) {
      parentIds = [
        typeof options.parent === 'string'
          ? options.parent
          : options.parent.id,
      ];
    }

    // Extract confidence from payload if present
    let confidence = options.confidence;
    const payloadCopy = { ...payload };
    if (confidence === undefined && 'confidence' in payloadCopy) {
      confidence = payloadCopy.confidence as number;
      delete payloadCopy.confidence;
    }

    // Create event
    await this.client!.events.create({
      scope_id: this.scope!.id,
      actor_id: resolvedActor.id,
      event_type_id: et.id,
      intent,
      payload: payloadCopy,
      confidence,
      parent_event_ids: parentIds,
      reason: options.reason,
    });

    // Wait for async processing and fetch event
    const shouldWait = options.wait !== false;
    if (shouldWait && this.waitTime > 0) {
      await sleep(this.waitTime);
      const events = await this.client!.events.list(this.scope!.id);
      for (const event of events) {
        if (
          event.intent === intent &&
          event.actor_id === resolvedActor.id
        ) {
          this.lastEvent = event;
          return event;
        }
      }
    }

    return null;
  }

  getClient(): LineageClient {
    this.ensureInitialized();
    return this.client!;
  }

  getScope(): Scope {
    this.ensureInitialized();
    return this.scope!;
  }

  getLastEvent(): Event | null {
    return this.lastEvent;
  }
}

export const globalState = new GlobalState();

File: sdks/python/src/lineage/_state.py
"""Global state management for the simple Lineage API."""

from __future__ import annotations

import time
from threading import Lock
from typing import Any
from uuid import UUID

from .client import LineageClient
from .enums import ActorType, Intent
from .models import Actor, Event, EventType, Scope


class _GlobalState:
    """Singleton holding global SDK state."""

    def __init__(self):
        self._lock = Lock()
        self._initialized = False
        self._client: LineageClient | None = None
        self._scope: Scope | None = None
        self._default_actor: Actor | None = None
        self._event_types: dict[str, EventType] = {}
        self._actors: dict[tuple[str, str], Actor] = {}  # (type, name) -> Actor
        self._last_event: Event | None = None
        self._wait_time: float = 2.0  # seconds to wait for async processing

    def init(
        self,
        project: str,
        *,
        base_url: str = "http://localhost:8080",
        domain: str | None = None,
        environment: str | None = None,
        actor_name: str | None = None,
        actor_type: str = "service",
        auto_create: bool = True,
        wait_time: float = 2.0,
    ) -> None:
        """Initialize the SDK with project and optional default actor."""
        with self._lock:
            self._client = LineageClient(base_url=base_url)
            self._wait_time = wait_time

            if auto_create:
                # Create or find scope
                self._scope = self._client.scopes.create(
                    project=project,
                    domain=domain,
                    environment=environment,
                )

                # Create default actor if specified
                if actor_name:
                    self._default_actor = self._get_or_create_actor(actor_type, actor_name)

            self._initialized = True

    def _ensure_initialized(self) -> None:
        """Raise error if not initialized."""
        if not self._initialized:
            raise RuntimeError("lineage.init() must be called first")

    def _get_or_create_actor(
        self, actor_type: str, actor_name: str, metadata: dict | None = None
    ) -> Actor:
        """Get or create an actor by type and name."""
        key = (actor_type, actor_name)
        if key in self._actors:
            return self._actors[key]

        actor = self._client.actors.create(
            type=ActorType(actor_type),
            name=actor_name,
            metadata=metadata,
        )
        self._actors[key] = actor
        return actor

    def _get_or_create_event_type(
        self, name: str, allowed_intents: list[Intent] | None = None
    ) -> EventType:
        """Get or create an event type by name."""
        if name in self._event_types:
            return self._event_types[name]

        # Default: allow all intents
        if allowed_intents is None:
            allowed_intents = list(Intent)

        event_type = self._client.event_types.create(
            name=name,
            version="1.0",
            allowed_intents=allowed_intents,
        )
        self._event_types[name] = event_type
        return event_type

    def _resolve_actor(
        self, actor: tuple[str, str] | Actor | None
    ) -> Actor:
        """Resolve actor from tuple (type, name), Actor object, or use default."""
        if actor is None:
            if self._default_actor is None:
                raise ValueError("No actor specified and no default actor set")
            return self._default_actor

        if isinstance(actor, Actor):
            return actor

        if isinstance(actor, tuple) and len(actor) == 2:
            return self._get_or_create_actor(actor[0], actor[1])

        raise ValueError(f"Invalid actor: {actor}")

    def emit(
        self,
        event_type: str,
        intent: str | Intent,
        payload: dict[str, Any],
        *,
        confidence: float | None = None,
        actor: tuple[str, str] | Actor | None = None,
        parent: Event | UUID | str | None = None,
        reason: str | None = None,
        wait: bool = True,
    ) -> Event | None:
        """Emit an event."""
        self._ensure_initialized()

        # Resolve intent
        if isinstance(intent, str):
            intent = Intent(intent)

        # Resolve actor
        resolved_actor = self._resolve_actor(actor)

        # Get or create event type
        et = self._get_or_create_event_type(event_type, [intent])

        # Resolve parent
        parent_ids = None
        if parent:
            if isinstance(parent, Event):
                parent_ids = [parent.id]
            elif isinstance(parent, (UUID, str)):
                parent_ids = [str(parent)]

        # Extract confidence from payload if present
        if confidence is None and isinstance(payload, dict):
            confidence = payload.pop("confidence", None)

        # Create event
        self._client.events.create(
            scope_id=self._scope.id,
            actor_id=resolved_actor.id,
            event_type_id=et.id,
            intent=intent,
            payload=payload,
            confidence=confidence,
            parent_event_ids=parent_ids,
            reason=reason,
        )

        # Wait for async processing and fetch event
        if wait and self._wait_time > 0:
            time.sleep(self._wait_time)
            events = self._client.events.list(scope_id=self._scope.id)
            for event in events:
                if (
                    event.intent == intent
                    and event.actor_id == resolved_actor.id
                ):
                    self._last_event = event
                    return event

        return None

    @property
    def client(self) -> LineageClient:
        """Get the underlying client for advanced operations."""
        self._ensure_initialized()
        return self._client

    @property
    def scope(self) -> Scope:
        """Get the current scope."""
        self._ensure_initialized()
        return self._scope

    @property
    def last_event(self) -> Event | None:
        """Get the last emitted event."""
        return self._last_event


# Global singleton
_global_state = _GlobalState()
